### PR TITLE
[BEAM-8335] Add the PTuple

### PR DESCRIPTION
**Pull request by rohdesamuel**
This PR adds the PTuple, which is analogous to Java's PCollectionTuple.

This is a very useful output type for composites that want to emit multiple PCollections where each have their own producer. This is a fundamentally different type and a more general type than the DoOutputsTuple. The DoOutputsTuple assumes a single producer for all of its PCollections; this does not.

This is being used in https://github.com/apache/beam/pull/9953.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by rohdesamuel** [comment]
R: @pabloem for review please and thank you!

**Comment by pabloem** [comment]
Can you show some examples of how this would be used? Would this be exposed to users at all? Java has a PCollectionTuple that is used in transforms like CoGBK, and Flatten. In Python we use a tuple containing PCollections e.g. `(pcoll1, pcoll2, pcoll3) | Flatten()`.

How would `PTuple` be used?

**Comment by rohdesamuel** [comment]
> Can you show some examples of how this would be used? Would this be exposed to users at all? Java has a PCollectionTuple that is used in transforms like CoGBK, and Flatten. In Python we use a tuple containing PCollections e.g. `(pcoll1, pcoll2, pcoll3) | Flatten()`.
> 
> How would `PTuple` be used?

`PTuple`s will be used like:
```
class SomeCustomTransform(PTransform):
  def expand(self, pcoll):
    def my_multi_do_fn(x):
      if isinstance(x, int):
        yield pvalue.TaggedOutput('number', x)
      if isinstance(x, str):
        yield pvalue.TaggedOutput('string', x)

  def printer(x):
    print(x)
    yield x

  outputs = pcoll | beam.ParDo(my_multi_do_fn).with_outputs()
  return pvalue.PTuple({
      'number': output.number | beam.ParDo(printer),
      'string': output.string | beam.ParDo(printer)
  })

p = beam.Pipeline()

main = p | SomeCustomTransform()

# Prints out each element that is a number.
numbers = main.number | beam.ParDo(...)

# Prints out each element that is a string.
strings = main.string | beam.ParDo(...)
```

This is a fundamentally different and necessary type than what is available in the SDK. It allows for referencing an arbitrary amount of PCollections by name each with a different producer.

**Comment by pabloem** [comment]
This is very similar to DoOutputsTuple, but DoOutputsTuple is quite a bit ideosyncratic, and not super reusable. I think this LGTM. But to be sure, I'll also request review from r: @aaltay

**Comment by rohdesamuel** [comment]
Right, it's similar but a much more powerful construct than the DoOutputsTuple. The DoOutputsTuple assumes a single producer for each PCollection. This allows for each PCollection to be an output of a different PTransform. You can have much more interesting composites using a PTuple instead of a DoOutputsTuple.

**Comment by aaltay** [comment]
@robertwb mentioned [1] that dictionaries can be returned as the output of composite transforms. Have you tried that?

[1] https://lists.apache.org/thread.html/519e95c5561466403ca3e2c1352d69415f8cb1753af345dfaaab7943@%3Cdev.beam.apache.org%3E

**Comment by robertwb** [comment]
Yes, I'd rather not introduce a new PTuple type given that Python already has first class support for tuples and dicts (and, for that matter, named tuples which, due to being a subclass of tuple, should just work). DoOutputsTuple are a bit special in that the set of outputs is dynamically determined.

**Comment by rohdesamuel** [comment]
I agree that I can use either dicts or tuples but I don't agree that they are a good user experience. Look at how you access the PCollection: dicts are accessed by strings, tuples are accessed by ints, namedtuples are accessed by field names, DoOutputsTuples are accessed by any of the prior. Now imagine that you have an arbitrary PTransform with multiple outputs, because of all the different access patterns the user won't know how to access the outputs. While it may be nice to have "first class support" in theory, in practice it doesn't work well.

That's where the PTuple comes in. This gives the user what they want: a homogeneous access pattern for multiple outputs.

File: sdks/python/apache_beam/pvalue.py [hunk 1]
```diff
@@ -201,6 +201,43 @@ class PDone(PValue):
...
+  composite.
+  """
+
+  def __init__(self, pcoll_dict):
+    """Initializes this named tuple with a dictionary of tagged PCollections.
+    """
+    self._pcolls = pcoll_dict
+
+  def __str__(self):
+    return '<%s>' % self._str_internal()
```
Comment (robertwb) [inline sdks/python/apache_beam/pvalue.py]: Remove and let it default to `__repr__`.

File: sdks/python/apache_beam/pvalue.py [hunk 2]
```diff
@@ -201,6 +201,43 @@ class PDone(PValue):
...
+
+  def __repr__(self):
+    return '<%s at %s>' % (self._str_internal(), hex(id(self)))
```
Comment (robertwb) [inline sdks/python/apache_beam/pvalue.py]: The representation should be PTuple(key1=pc1, key2=pc2, ...)

File: sdks/python/apache_beam/pvalue.py [hunk 3]
```diff
@@ -201,6 +201,43 @@ class PDone(PValue):
...
+  def _str_internal(self):
+    return '%s pcollections=%s' % (
+        self.__class__.__name__, self._pcolls)
+
+  def __iter__(self):
+    for tag in self._pcolls:
+      yield self[tag]
+
+  def __getattr__(self, tag):
+    # Special methods which may be accessed before the object is
```
Comment (robertwb) [inline sdks/python/apache_beam/pvalue.py]: We shouldn't support pickling these.

Comment (rohdesamuel) [reply sdks/python/apache_beam/pvalue.py]: Removed

File: sdks/python/apache_beam/pvalue.py (hunk 2 above)
Comment (rohdesamuel) [reply sdks/python/apache_beam/pvalue.py]: Ack, done.

File: sdks/python/apache_beam/pvalue.py (hunk 1 above)
Comment (rohdesamuel) [reply sdks/python/apache_beam/pvalue.py]: done

**Pull request closed without merging**